src/external/wannier_shared.jl

```diff
@@ -0,0 +1,407 @@
+# Shared functionality to operate with Wannier programs (Wannier.jl and Wannier90).
+# The communication happens with the Wannier90 file formats which Wannier.jl also understands.
+
+using Dates
+
+@doc raw"""
+Base type for Wannier projections.
+
+Wannierization searchs for a unitary matrix ``U_{m_n}``.
+As a starting point for the search, we can provide an initial guess function ``g``
+for the shape of the Wannier functions, based on what we expect from knowledge of the problem or physical intuition.
+This starting matrix is called ``[A_k]_{m,n}``, and is computed as follows:
+``[A_k]_{m,n} = \langle ψ_m^k | g^{\text{per}}_n \rangle``.
+The matrix will be orthonormalized by the chosen Wannier program, we don't need to do so ourselves.
+"""
+abstract type WannierProjection end
+
+# Dispatch function for WannierProjection implementations, evaluating g in Fourier space.
+function get_fourier_projection_coefficients end
+
+"""
+A Gaussian-shaped initial guess. Can be used as an approximation of an s- or σ-like orbital.
+"""
+struct GaussianWannierProjection <: WannierProjection
+    center::AbstractVector{Float64}
+end
+
+function get_fourier_projection_coefficients(proj::GaussianWannierProjection, basis::PlaneWaveBasis, qs)
+    # associate a center with the fourier transform of the corresponding gaussian
+    # TODO: what is the normalization here?
+
+    model = basis.model
+    [begin
+        q_cart = recip_vector_red_to_cart(model, q)
+        exp( 2π*(-im*dot(q, proj.center) - dot(q_cart, q_cart) / 4) )
+    end for q in qs]
+end
+
+function random_gaussian_projection()
+    GaussianWannierProjection(rand(3))
+end
+
+"""
+A sum of two Gaussians. Can be used as an approximation of a p-like orbital.
+"""
+struct DoubleGaussianWannierProjection <: WannierProjection
```

Comment:
And here is a prime example why being able to just use functions is great. Then the 40 lines of code to make the rather restrictive type `DoubleGaussianWannierProjection` can be reduced to just 5:

```
function plike_gaussian_projection(center1::AbstractVector, center2::AbstractVector)
    function inner(basis::PlaneWaveBasis, qs)
       GaussianWannierProjection(center1)(basis, qs) - GaussianWannierProjection(center1)(basis, qs)
    end
end
```